File: src/main/java/org/nia/logic/quests/buyfish/SellFishQuest.java
package org.nia.logic.quests.buyfish;

import org.nia.logic.quests.IQuest;
import org.nia.logic.quests.IQuestEvent;
import org.nia.logic.quests.IQuestStep;

import java.util.Arrays;
import java.util.List;
import java.util.Random;

/**
 * @author IANazarov
 */
public class SellFishQuest implements IQuest {
    public static SellFishQuest INSTANCE = new SellFishQuest();

    @Override
    public String getStart() {
        return "Остап попросил тебя поторговать рыбой на рынке. Ему недавно привезли несколько повозок с рыбой, и " +
                "теперь её надо продать, пока не испортилась.";
    }

    private enum SellFishEvent implements IQuestEvent {
        SELL_FISH_NOISY_MAN(NoisyMan.INIT) {
            @Override
            public IQuestStep getQuestStep(String questStep) {
                return NoisyMan.valueOf(questStep);
            }
        },
        SELL_FISH_INSPECTION(Inspection.INIT) {
            @Override
            public IQuestStep getQuestStep(String questStep) {
                return Inspection.valueOf(questStep);
            }
        },
        SELL_FISH_TURTLE_PANDA(TurtlePanda.INIT) {
            @Override
            public IQuestStep getQuestStep(String questStep) {
                return TurtlePanda.valueOf(questStep);
            }
        },
        SELL_FISH_MAGIC_FISH(MagicFish.INIT) {
            @Override
            public IQuestStep getQuestStep(String questStep) {
                return MagicFish.valueOf(questStep);
            }
        },
        SELL_FISH_HOBBIT(DragonMeat.INIT) {
            @Override
            public IQuestStep getQuestStep(String questStep) {
                return DragonMeat.valueOf(questStep);
            }
        };
        private IQuestStep init;

        SellFishEvent(IQuestStep init) {
            this.init = init;
        }

        @Override
        public IQuestStep getInit() {
            return init;
        }

        @Override
        public String getName() {
            return name();
        }
    }

    @Override
    public IQuestEvent getRandomEvent() {
        List<SellFishEvent> events = Arrays.asList(SellFishEvent.values());
        return events.get(new Random().nextInt(events.size()));
    }

    @Override
    public IQuestEvent getEvent(String event) {
        return SellFishEvent.valueOf(event);
    }
}

File: src/main/java/ru/nia/tavern/quests/buyfish/SellFishQuest.kt
package ru.nia.tavern.quests.buyfish

import ru.nia.tavern.quests.IQuest
import ru.nia.tavern.quests.IQuestEvent
import ru.nia.tavern.quests.IQuestStep
import ru.nia.tavern.quests.QuestsEnum
import java.util.*

/**
 * @author IANazarov
 */
class SellFishQuest : IQuest {

    override val start: String
        get() = "Остап попросил тебя поторговать рыбой на рынке. Ему недавно привезли несколько повозок с рыбой, и " + "теперь её надо продать, пока не испортилась."

    override val randomEvent: IQuestEvent
        get() {
            val events = Arrays.asList(*SellFishEvent.values())
            return events[Random().nextInt(events.size)]
        }

    enum class SellFishEvent private constructor(override val init: IQuestStep) : IQuestEvent {
        SELL_FISH_NOISY_MAN(NoisyMan.INIT) {
            override fun getQuestStep(questStep: String): IQuestStep {
                return NoisyMan.valueOf(questStep)
            }
        },
        SELL_FISH_INSPECTION(Inspection.INIT) {
            override fun getQuestStep(questStep: String): IQuestStep {
                return Inspection.valueOf(questStep)
            }
        },
        SELL_FISH_TURTLE_PANDA(TurtlePanda.INIT) {
            override fun getQuestStep(questStep: String): IQuestStep {
                return TurtlePanda.valueOf(questStep)
            }
        },
        SELL_FISH_MAGIC_FISH(MagicFish.INIT) {
            override fun getQuestStep(questStep: String): IQuestStep {
                return MagicFish.valueOf(questStep)
            }
        },
        SELL_FISH_HOBBIT(DragonMeat.INIT) {
            override fun getQuestStep(questStep: String): IQuestStep {
                return DragonMeat.valueOf(questStep)
            }
        };

        override val name: String
            get() = name

        override val questsEnum: QuestsEnum
            get() = QuestsEnum.SELL_FISH
    }

    override fun getEvent(event: String): IQuestEvent {
        return SellFishEvent.valueOf(event)
    }

    companion object {
        var INSTANCE = SellFishQuest()
    }
}
